### ISSUE TITLE
New flag for picture overwrite

### ISSUE THREAD

**Issue by panzli** [TLINDEN/kleingebaeck#39]
Hello Tom,

like mentioned in https://github.com/TLINDEN/kleingebaeck/issues/30#issuecomment-1894498469 it would be really nice, if there is an option to skip the picture download for already downloaded items.

But this feature has a very low priority...

Regards

**Comment by TLINDEN** [comment]
Could be done, but you are aware, that the images which are stored in a backup are not the originals anyway? And while I did not test it I think they won't touch uploaded images which already perfectly match their specs (as they do, since we downloaded them earlier).

Another problem is that it is next to impossible to tell which image is which. Of course they do have names, but these are just some non-sensical numbers. The tool renames them because of this reason and because it is hard to maintain files on a unix system with `$` characters in them:
```
https://img.kleinanzeigen.de/api/v1/prod-ads/images/60/606de6a0-b80b-4ae7-8ccf-5040bc8bb28e?rule=$_59.JPG
```

Of course I could create another naming scheme, but the problem would remain: how to check if precisely *this* images hasn't changed? Normally you'd check using a HTTP HEAD request. I'll take a look and see if kleinanzeigen.de provides an etag or something which would help to implement some caching strategy. 

Otherwise I can just go after the file names. But imagine you upload another image to an ad and re-order the images. Then what was `1.jpg` before is the image which was `4.jpg` now for example. When using such a scheme to avoid re-download images you might as well miss images (unless of course if you don't have the original image anymore).

So, IMHO, you should keep the original images somewhere in case you want to re-upload the same ones again.

**Comment by TLINDEN** [comment]
Ok, I'll use the [goimagehash](https://github.com/corona10/goimagehash) to compute a dhash of the downloaded image, compare that to existing ones regardless of their filename and will then not overwrite if the distance is very low, say < 5 or something.

The test shows good results. Here are 2 images, one original and a heavily manipulated copy. The caluclated distance between the two is just `1`.

![manip](https://github.com/TLINDEN/kleingebaeck/assets/1322812/a939c508-ae9d-4409-a647-0057c35565da)
![DSC_4391](https://github.com/TLINDEN/kleingebaeck/assets/1322812/20fa6b4f-cbf9-4615-8404-58ac9d6d75d0)

The only downside is, that I need to load all images on startup and pre-compute their dhashes. Maybe I'll ad an option to enable the feature or the other way around, disable it - whatever will be the default.

**Comment by TLINDEN** [comment]
Implementation is there, needs some polishing though.

Todo:
- add unit tests
- add flag to override the feature
- tune debug output

**Comment by TLINDEN** [comment]
@panzli the develop branch now contains the new code, which checks if an almost similar image already exists in the ad directory and skips it if this is the case.

I tried it with this ad: https://www.kleinanzeigen.de/s-anzeige/olympus-pen-e-pl1-kit-mit-mzuiko-14-42mm-objektiv-und-ledertasche/2653981706-245-19951. I removed all previously downloaded images and put the original ones in there. This is how the directory looks after the backup run:
```default
-rw-r--r--  1 scip scip   74015 Jan 22 14:45 11.jpg
-rw-r--r--  1 scip scip   43719 Jan 22 14:45 12.jpg
-rw-r--r--  1 scip scip   44009 Jan 22 14:45 6.jpg
-rw-r--r--  1 scip scip    1247 Jan 22 14:45 Adlisting.txt
-rw-r--r--  1 scip scip 4847990 Jan 22 14:44 DSC_4390.JPG
-rw-r--r--  1 scip scip 6273995 Jan 22 14:44 DSC_4391.JPG
-rw-r--r--  1 scip scip 6106485 Jan 22 14:44 DSC_4392.JPG
-rw-r--r--  1 scip scip 5590061 Jan 22 14:44 DSC_4393.JPG
-rw-r--r--  1 scip scip 5795861 Jan 22 14:44 DSC_4394.JPG
-rw-r--r--  1 scip scip 5243768 Jan 22 14:44 DSC_4395.JPG
-rw-r--r--  1 scip scip 6514017 Jan 22 14:44 DSC_4396.JPG
-rw-r--r--  1 scip scip 7174799 Jan 22 14:44 DSC_4397.JPG
-rw-r--r--  1 scip scip 6973552 Jan 22 14:44 DSC_4398.JPG
-rw-r--r--  1 scip scip 5887938 Jan 22 14:44 DSC_4399.JPG
-rw-r--r--  1 scip scip 6035717 Jan 22 14:44 DSC_4400.JPG
-rw-r--r--  1 scip scip 6849578 Jan 22 14:44 DSC_4401.JPG
-rw-r--r--  1 scip scip 4999896 Jan 22 14:44 DSC_4402.JPG
-rw-r--r--  1 scip scip 4793061 Jan 22 14:44 DSC_4403.JPG
```

So, 3 images have been downloaded, these are the ones I rotated in the kleinanzeigen backend. Otherwise it works like a charm. You may also look at the debug output of that run: https://gist.github.com/TLINDEN/69fc45a0678e057f592bf828207e3629.

So if you have the time and interest it would be great if you could test the new code as well.

---

note to self: missing TODOs still:
- use goroutines to speed up? Measure! (or not? It's good enough for Fehmarn as it is)
- check bytes.Buffer rewind problem, try to avoid copying of img data twice

**Comment by TLINDEN** [comment]
Ah my goodness, I implemented the feature but forgot to add the actual flag! Thanks for the hint!

**Comment by TLINDEN** [comment]
Ok, the flag is there now, also I closed the last TODO and decided to not use goroutines for the prefetcher.

So, the summary of all those changes:

- Images are now checked against the ones which eventually already exist in the ad directory of each ad (e.g. from a previous run or because the user copied the unaltered original files in there)
- IF there are images, they will be read, cached and a difference hash will be made of each. See [go module goimagehash](https://github.com/corona10/goimagehash) and if you need more detail, see [description of how the difference hashing works](https://www.hackerfactor.com/blog/index.php?/archives/529-Kind-of-Like-That.html)
- Then images are being downloaded, a difference hash is being computed for each and then the distance between each of the cached hashes of the existing images is being calculated. If the distance is below 5 we consider the image too similar to the existing one and don't overwrite it.
- This works for images which have been resized, cropped, color manipulated etc as long as they still show the original motive.
- Filenames are not considered.
- The whole behavior can be turnedd off using the `--force` flag.

Besides we also have this behavior regarding HTTP which might help as well:
- There's a flag `--ignoreerrors` when set causes the tool not to bail out if something goeas wrong during image download. Only HTTP errors are considered though, I/O errors still lead to certain dead.
- The HTTP code in kleingebäck uses its own Transport with retries, that is, each request will be repeated three time if it fails (e.g. with HTTP 503 etc), the loop will wait for an escalating time til if finally gives up. So, minor network glitches might be circumvented eventually.
- We also introduced support for Cookies, that is, during the very first time we hit kleinanzeigen.de all cookies will be stored and be used for each subsequent request. That way we show a more natural non-bot-like behavior.
- Image downloads happens using goroutines in parallel, but each download is being throttled a little bit (couple of milliseconds), so that the downloads to not hit the site simultaneously.
